gax/src/main/java/com/google/api/gax/rpc/ErrorDetails.java

```diff
@@ -0,0 +1,163 @@
+public abstract class ErrorDetails {
+
+  /**
+   * This is the most important and special error message. It describes the cause of the error with
+   * structured details that both humans and applications can depend on.
+   */
+  @Nullable
+  public ErrorInfo errorInfo() {
+    return unpack(ErrorInfo.class);
+  }
+
+  /**
+   * Describes when the clients can retry a failed request. Clients could ignore the recommendation
+   * here or retry when this information is missing from error responses.
+   */
+  @Nullable
+  public RetryInfo retryInfo() {
+    return unpack(RetryInfo.class);
+  }
+
+  /** Describes additional debugging info. */
+  @Nullable
+  public DebugInfo debugInfo() {
+    return unpack(DebugInfo.class);
+  }
+
+  /** Describes how a quota check failed. */
+  @Nullable
+  public QuotaFailure quotaFailure() {
+    return unpack(QuotaFailure.class);
+  }
+
+  /** Describes what preconditions have failed. */
+  @Nullable
+  public PreconditionFailure preconditionFailure() {
+    return unpack(PreconditionFailure.class);
+  }
+
+  /**
+   * Describes violations in a client request. This error type focuses on the syntactic aspects of
+   * the request.
+   */
+  @Nullable
+  public BadRequest badRequest() {
+    return unpack(BadRequest.class);
+  }
+
+  /**
+   * Contains metadata about the request that clients can attach when filing a bug or providing
+   * other forms of feedback.
+   */
+  @Nullable
+  public RequestInfo requestInfo() {
+    return unpack(RequestInfo.class);
+  }
+
+  /** Describes the resource that is being accessed. */
+  @Nullable
+  public ResourceInfo resourceInfo() {
+    return unpack(ResourceInfo.class);
+  }
+
+  /** Provides links to documentation or for performing an out-of-band action. */
+  @Nullable
+  public Help help() {
+    return unpack(Help.class);
+  }
+
+  /**
+   * Provides a localized error message that is safe to return to the user which can be attached to
+   * an RPC error
+   */
+  @Nullable
+  public LocalizedMessage localizedMessage() {
+    return unpack(LocalizedMessage.class);
+  }
+
+  /** This is a list of raw/unparsed error messages that returns from server. */
+  @Nullable
+  abstract List<Any> rawErrorMessages();
+
+  public static Builder builder() {
+    return new AutoValue_ErrorDetails.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder setRawErrorMessages(List<Any> rawErrorMessages);
+
+    public abstract ErrorDetails build();
+  }
+
+  @VisibleForTesting
+  <T extends Message> T unpack(Class<T> errorTypeClazz) {
+    List<Any> rawErrorMessages = rawErrorMessages();
+    if (rawErrorMessages == null) {
+      return null;
+    }
+    for (Any detail : rawErrorMessages) {
+      if (!detail.is(errorTypeClazz)) {
+        continue;
+      }
+      try {
+        return detail.unpack(errorTypeClazz);
+      } catch (InvalidProtocolBufferException e) {
+        throw new ProtocolBufferParsingException(e);
```

Comment:
Can we add a message to the exception saying something like "Failed to unpack ${class} from error details"?

Reply:
I gave it some thoughts as well, I decided not to add an additional message because the message will be defaulting to the message of cause, which I think might be better. Something like this 
```
com.google.protobuf.InvalidProtocolBufferException: Protocol message end-group tag did not match expected tag.
com.google.api.gax.rpc.ProtocolBufferParsingException: 
	at com.google.api.gax.rpc.ErrorDetails.unpack(ErrorDetails.java:158)
```
compare to 
```
Failed to unpack ErrorInfo from error details
com.google.api.gax.rpc.ProtocolBufferParsingException: 
	at com.google.api.gax.rpc.ErrorDetails.unpack(ErrorDetails.java:158)
```
The latter might be easier to understand but the former contains more information, even though end users can not do much about it, but might be useful for troubleshooting/reporting issues.
Let me know if this makes sense or not.

Reply:
The messages should always be preserved in the causing exception. You would see it in the stack trace anyway. So, I'm not sure what you mean by "former contains more information".

Reply:
Yes, I agree. I should probably say "more information on the top of the stacktrace", which I feel like might be better/good enough. But I can definitely add a message if we think it provides more clarity.

Reply:
Added.